File: Client/Desktop/src/main/java/com/example/tin/Connection.java
package com.example.tin;

import java.io.*;
import java.net.Socket;
import java.nio.ByteBuffer;
import java.nio.CharBuffer;
import java.nio.charset.StandardCharsets;
import java.sql.SQLOutput;

public class Connection {

    String serverAddress = "192.168.0.95";
    int portNumber = 9995;
    Socket socket;
    DataOutputStream outToServer;
    DataInputStream inFromServer;
    boolean isConnected = false;


    private void changePort(int port) {
        portNumber = port;
        connect();
    }

    public boolean connect() {
        if (!isConnected){
            try {
                System.out.println("Lacze z serwerem...");
                socket = new Socket(serverAddress, portNumber);
                //socket.connect(new InetSocketAddress(serverAddress, portNumber));
                System.out.println("Polaczono");
                isConnected = true;
                outToServer = new DataOutputStream(socket.getOutputStream());
                inFromServer = new DataInputStream(socket.getInputStream());
                return true;
            } catch (Exception ex) {
                System.out.println("Unable to connect with server!");
                isConnected = false;
                return false;
            }
        }
        return true;
    }

    public void disconnect() {
        try {
            socket.close();
        } catch (IOException ex) {
            System.out.println("Problem with disconnecting");
        }
    }

    public int sendMessage(String message) throws IOException {

        System.out.println("Pisze info");
        System.out.println(message);
        outToServer.writeBytes(message);
        System.out.println("Napisalem info");
        return 0;
    }

    public String receiveMessage() throws IOException {

        try {
            System.out.println("Czekam na wiadomosc");
            ByteBuffer size = ByteBuffer.allocate(4);
            inFromServer.readFully(size.array(), 0, 4);
            ByteBuffer buffer = ByteBuffer.allocate(Integer.parseInt(new String(size.array(), StandardCharsets.UTF_8)));
            inFromServer.readFully(buffer.array());
            String message = new String(buffer.array());
            System.out.println("Wiadomosc: " + message);
            return message;
        } catch (IOException ex) {
            disconnect();
            System.out.println("Problem with receiving message! Disconnecting...");
            return "";
        }
    }

    public boolean getConnected() {
        return isConnected;
    }
}

File: Client/Android/app/src/main/java/com/example/tin/network/Connection.kt
package com.example.tin.network

import java.io.DataInputStream
import java.io.IOException
import java.io.OutputStreamWriter
import java.net.Socket
import java.net.UnknownHostException
import java.nio.charset.StandardCharsets

object Connection : Runnable {

    interface NetworkEventListener {
        fun incomingMessage(message: String)
    }

    private var serverAddress = "192.168.0.11"
    private var portNumber = 9999
    private lateinit var socket: Socket
    private lateinit var outToServer: OutputStreamWriter
    private lateinit var inFromServer: DataInputStream
    private var isRunning = true
    private var isConnected = false
    private var networkEventListener: NetworkEventListener? = null
    private var shouldReconnect = true
    private var lock = Object()

    fun setNetworkEventListener(networkEventListener: NetworkEventListener) {
        this.networkEventListener = networkEventListener
    }

    init {

    }

    private fun tryConnect(): Boolean {
        try {
            socket = Socket(serverAddress, portNumber)
            outToServer = OutputStreamWriter(socket.getOutputStream())
            inFromServer = DataInputStream(socket.getInputStream())
        } catch (e: UnknownHostException) {
            return false
        } catch (e: IOException) {
            return false
        }
        return true
    }

    private fun connect() {
        while (!tryConnect());
        isConnected = true
    }

    @Synchronized
    override fun run() {
        connect()

        while (isRunning) {
            if (!isConnected) {
                if (shouldReconnect) {
                    connect()
                } else {
                    synchronized(lock) {
                        lock.wait()
                    }
                }
            }
            try {
                val readMessage = readMessage()
                try {
                    networkEventListener!!.incomingMessage(readMessage)
                } catch (e: NullPointerException) {

                }
            } catch (e: IOException) {
                print(e)
            }
        }
    }

    private fun readMessage(): String {
        if (!isConnected) {
            connect()
        }
        try {
            val messageSize = Integer.parseInt(read(4))
            return read(messageSize)
        } catch (e: IOException) {
            throw e
        } catch (e: NumberFormatException) {
            throw IOException()
        }
    }

    private fun read(bytesToRead: Int): String {
        val messageBuffer = ByteArray(bytesToRead)
        var bytesRead = 0
        try {
            while (bytesRead != bytesToRead) {
                val result = inFromServer.read(messageBuffer, bytesRead, bytesToRead - bytesRead)
                if (result == -1) {
                    isConnected = false
                    throw IOException()
                }
                bytesRead += result
            }
        } catch (e: IOException) {
            throw e
        }
        return String(messageBuffer, StandardCharsets.UTF_8)
    }

    fun disconnect() {
        isConnected = false
        shouldReconnect = false
        socket.close()
    }

    fun isConnected(): Boolean {
        return isConnected
    }

    fun sendMessage(message: String) {
        if (!isConnected) {
            connect()
        }
        val seizedMessage = String.format("%04d$message", message.length)
        outToServer.write(seizedMessage)
        outToServer.flush()
    }

    fun reconnect() {
        shouldReconnect = true
        synchronized(lock) {
            lock.notify()
        }
    }
}